### DISCUSSION TITLE
Ruby 3.2 no longer bundling source for libffi

### DISCUSSION THREAD

**Discussion by AndyObtiva** [kojix2/LibUI#67] [General]
I just read this in the Ruby 3.2 preview release notes:

https://www.ruby-lang.org/en/news/2022/11/11/ruby-3-2-0-preview3-released/

> No longer bundle 3rd party sources
> We no longer bundle 3rd party sources like libyaml, libffi.
> 
> libyaml source has been removed from psych. You may need to install libyaml-dev with Ubuntu/Debian platfrom. The package name is different each platforms.
> 
> bundled libffi source is also removed from fiddle

Is this going to be an issue for us in ensuring people don’t have to install anything other than LibUI + Glimmer DSL for LibUI to build GUI ?

If so, should you be asking the Ruby team to keep bundling it and explain why that’s important?

Here is the Ruby discussion about it: 
https://bugs.ruby-lang.org/issues/18571

Perhaps I am misunderstanding this. It seems when building Ruby with something like RVM, libffi should now be installed as part of that. And, if so, then it is still handled automatically and users won’t need to do anything themselves. 

If it won’t be an issue, then you may disregard my comment and we can close this discussion.

**Comment by rubyFeedback** [comment]
Hmm. For my linux build I compile everything from source (using ruby) so I don't think it will be an issue for me.

For windows, though, I think this depends. Will the one-click installer include these? I think if so then it should make no difference.

I don't know the situation on OSX.

I have not thought this through, but I'll make a prediction and assume it will not cause an issue. But it's good that this is raised to the ruby core team before the xmas release of the next ruby. During xmas time I intend to test libui-ng and I'll also try to get into glimmer + libui; I even have a good use case (traditional pen and paper roleplaying games, I want to see how easy it is to use glimmer libui for it; my only concern is with the "pure DSL approach versus modular include-like integration", but I may try both variants and then see what may be easier to do, solely for windows systems).

**Comment by kojix2** [comment]
Thank you for the information.
I am not sure, I will need to look into it further.

**Comment by AndyObtiva** [comment]
Ok thank you for the clarification. We can close this discussion.

**Comment by AndyObtiva** [comment]
Sorry, I reopened the discussion because I just noticed this statement (I somehow missed it the first time I read the reply):

> "pure DSL approach versus modular include-like integration"

There is no "versus" when using Glimmer DSL for LibUI. You actually `include Glimmer::LibUI::Application`, `include Glimmer::LibUI::CustomControl`, or include any custom modules you want to include, and thus are able to do "include-like integration" WITH the DSL approach. It is not "pure DSL" in the sense of "DSL-Only". It is a mix-and-match-approach to solving problems that allows you to use every possible technique available in Ruby, but with less code. In other words, the only thing different about the "DSL approach" is that you write LESS CODE than a "Non-DSL approach", meaning you AUTOMATE the solution with fewer KEYWORDS. You should not need to "try both variants" because a non-DSL approach would just be you writing 4x-10x the amount of code or more without gaining any benefits (meaning it’s a waste of time). It's better to just write your solution in Glimmer DSL for LibUI, and if there is anything you want to see improved in it, report it and have it improved.

Cheers!

**Comment by kojix2** [comment]
I asked about the libffi code being removed from Ruby on the [ruby-jp](https://ruby-jp.github.io/) slack and received answers from two Ruby core developers hsbt and mame. I also received an answer from my mentor kou.

To summarize their response, they say that you can build Ruby without Fiddle by adding the following options when installing Ruby from source code.
`. /configure --without-ext=+,fiddle` 
Without such an option, the build fails. So, in fact, there is little chance that there is no Fiddle.
If you use tools such as RVM or rbenv to install Ruby, it seems safe to assume that Fiddle is available.

However, as shevy said, problems can occur on Windows. I think Fiddle will be packaged with RubyInstaller, but we have to wait for the RubyInstaller release to see if it actually works. We have had problems with Windows and Fiddle before.  But that was fixed by the community.
